File: new/bot/handlers/start.py
from aiogram import Router, types, F
from aiogram.filters import CommandStart, Command
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from core.models import User, Referral
from core.config import settings
import time

router = Router()

async def check_channels(bot, user_id):
    for channel in settings.channel_list:
        try:
            member = await bot.get_chat_member(channel, user_id)
            if member.status in ["left", "kicked"]:
                return False
        except Exception:
            return False
    return True

@router.message(CommandStart())
async def cmd_start(message: types.Message, db: AsyncSession, bot):
    user_id = message.from_user.id
    args = message.text.split()[1:] if len(message.text.split()) > 1 else None
    
    # 1. Check/Register User
    result = await db.execute(select(User).where(User.telegram_id == user_id))
    user = result.scalar_one_or_none()
    
    is_new = False
    if not user:
        is_new = True
        referrer_id = int(args[0]) if args and args[0].isdigit() else None
        user = User(
            telegram_id=user_id,
            username=message.from_user.username,
            first_name=message.from_user.first_name,
            referrer_id=referrer_id
        )
        db.add(user)
        await db.commit()
    
    # 2. Check Channels
    subscribed = await check_channels(bot, user_id)
    
    if not subscribed:
        kb = types.InlineKeyboardMarkup(inline_keyboard=[
            [types.InlineKeyboardButton(text="Join Channel", url=f"https://t.me/{c.replace('@','')}") for c in settings.channel_list],
            [types.InlineKeyboardButton(text="✅ Check Subscription", callback_data="check_sub")]
        ])
        return await message.answer("Please join our channels to participate in the contest:", reply_markup=kb)

    # 3. Handle Referral Points (once)
    if is_new and user.referrer_id and subscribed:
        # Check daily limit (Simplified for now)
        from sqlalchemy import func
        today = types.DateTime.now().date()
        ref_count_res = await db.execute(select(func.count(Referral.id)).where(Referral.referrer_id == user.referrer_id))
        if ref_count_res.scalar() < 10:
            ref_user_res = await db.execute(select(User).where(User.telegram_id == user.referrer_id))
            ref_user = ref_user_res.scalar_one_or_none()
            if ref_user:
                ref_user.score += 50
                db.add(Referral(referrer_id=user.referrer_id, referred_user_id=user_id))
                await db.commit()
                try:
                    await bot.send_message(user.referrer_id, f"🎉 New referral! You earned 50 pts from {message.from_user.first_name}")
                except: pass

    await message.answer(f"Welcome to the Contest Platform, {message.from_user.first_name}! 🚀\nYour current score: {user.score} pts", reply_markup=main_menu_kb())

def main_menu_kb():
    return types.ReplyKeyboardMarkup(keyboard=[
        [types.KeyboardButton(text="🎁 Daily Check-in"), types.KeyboardButton(text="🧩 Solve Quiz")],
        [types.KeyboardButton(text="🏆 Leaderboard"), types.KeyboardButton(text="👥 Referrals")],
        [types.KeyboardButton(text="📱 Open Web App", web_app=types.WebAppInfo(url=settings.TELEGRAM_WEBHOOK_URL.replace('/bot/webhook.php', '/web/index')))]
    ], resize_keyboard=True)

File: bot/commands/start.php
<?php
// bot/commands/start.php

function handleStart($message, $telegram, $db)
{
    $chatId = $message['chat']['id'];
    $from = $message['from'];
    $text = $message['text'];
    $username = $from['username'] ?? null;
    $firstName = $from['first_name'] ?? 'User';

    // 1. Check if user exists
    $user = $db->fetch("SELECT * FROM users WHERE telegram_id = :tid", [':tid' => $chatId]);

    $isNewUser = false;
    if (!$user) {
        $isNewUser = true;
        // Parse referral ID
        $referrerId = null;
        if (preg_match('/\/start ref_(\d+)/', $text, $matches)) {
            $referrerId = $matches[1];
        }

        // Save new user
        $db->query("INSERT INTO users (telegram_id, username, first_name, referrer_id) VALUES (:tid, :un, :fn, :rid)", [
            ':tid' => $chatId,
            ':un' => $username,
            ':fn' => $firstName,
            ':rid' => $referrerId
        ]);

        $user = $db->fetch("SELECT * FROM users WHERE telegram_id = :tid", [':tid' => $chatId]);
    }

    // 2. Check channel membership
    $config = require __DIR__ . '/../../config/telegram.php';
    $notJoined = [];
    foreach ($config['channel_ids'] as $channel) {
        if (!$telegram->isJoined($chatId, $channel)) {
            $notJoined[] = $channel;
        }
    }

    if (!empty($notJoined)) {
        $keyboard = ['inline_keyboard' => []];
        foreach ($notJoined as $channel) {
            $keyboard['inline_keyboard'][] = [['text' => "Join $channel", 'url' => "https://t.me/" . str_replace('@', '', $channel)]];
        }
        $keyboard['inline_keyboard'][] = [['text' => "Check Membership", 'callback_data' => "check_membership"]];

        $telegram->sendMessage($chatId, "Welcome! To participate in the contest, please join our channels first:", $keyboard);
        return;
    }

    // 3. If new user joined channels, add referral points to referrer
    if ($isNewUser && $user['referrer_id']) {
        // Anti-cheat: Check daily limit
        $today = date('Y-m-d');
        $refCount = $db->fetch("SELECT COUNT(*) as count FROM referrals WHERE referrer_id = :rid AND DATE(created_at) = :today", [
            ':rid' => $user['referrer_id'],
            ':today' => $today
        ]);

        if ($refCount['count'] < 10) { // Default limit from 13-Anti_Cheat_System.md
            $db->query("INSERT INTO referrals (referrer_id, referred_user_id) VALUES (:rid, :ruid)", [
                ':rid' => $user['referrer_id'],
                ':ruid' => $chatId
            ]);
            $db->query("UPDATE users SET score = score + 50 WHERE telegram_id = :rid", [':rid' => $user['referrer_id']]);
            $telegram->sendMessage($user['referrer_id'], "🎉 Someone joined using your link! You earned 50 points.");
        }
    }

    // 4. Show Main Menu
    showMainMenu($chatId, $telegram, $user);
}

function showMainMenu($chatId, $telegram, $user)
{
    $text = "Hello, {$user['first_name']}!\n\n";
    $text .= "🏆 Your Score: {$user['score']}\n";
    $text .= "🔥 Level: {$user['level']}\n\n";
    $text .= "Use the menu below to earn points:";

    $keyboard = [
        'keyboard' => [
            [['text' => "📅 Daily Check-in"], ['text' => "🧠 Quiz"]],
            [['text' => "📊 Leaderboard"], ['text' => "👥 Referrals"]],
            [['text' => "🎁 Redeem Code"], ['text' => "📱 Open Web Panel"]]
        ],
        'resize_keyboard' => true
    ];

    $telegram->sendMessage($chatId, $text, $keyboard);
}
